### PR TITLE
Add some more bijections

### PR DESCRIPTION
**Pull request by aseyboldt**
This adds

- DCT: discrete cosine transform
- AsymmetricAffine: a smooth nonlinear transform that applies different scaling factors for positive and negative values
- Householder: a householder transformation
- Neg: multiply by -1
- Sandwich: apply a transform, then a second and then the inverse of the first

### CODE REVIEW

File: flowjax/bijections/orthogonal.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    """
+    shape: tuple[int, ...]
+    params: Array
+    cond_shape = None
+
+    def __init__(self, params: Array):
+        self.shape = (params.shape[-1],)
+        self.params = params
+
+    def _householder(self, x: Array, params: Array) -> Array:
```
Comment (danielward27) [inline flowjax/bijections/orthogonal.py:56]: Looks like we can simplify to
```
vec = params / jnp.linalg.norm(params)
```

**Review by danielward27 — CHANGES_REQUESTED** [review]
Cheers! I have been meaning to add the ``Householder`` transform for a while! I had a little play with the discrete cosine transform too - it definitely gives some fun possibilities for modelling sequence/time series data.

I do wan't to be a bit careful to avoid overly expanding the number of bijections included in FlowJAX, so I think omitting ``Neg`` and ``AsymmetricAffine`` is probably sensible for now. I've added some comments, and if you are happy to incorporate the changes I'll merge it in. Thanks for the pull request!

File: flowjax/bijections/orthogonal.py (hunk 1 above)
Comment (danielward27) [reply flowjax/bijections/orthogonal.py:56]: Maybe a more readable option, is to instead store:
```python
self.unit_vec = paramax.Parameterize(lambda x: x/jnp.linalg.norm(x), params)
```
and use:
```python
x - 2 * self.unit_vec * (x @ self.unit_vec)
```

**Comment by aseyboldt** [comment]
Thanks for the review, I'll fix those early next week.

About AsymmetricAffine: That one doesn't have a reference, I made it up during experimentation. I wanted to have something between an affine and the rational quadratic spline as transformation in the coupling layer. For some posteriors that seems to work very well. Wasn't easy to come up with something where neither the forward nor inverse blows up, and that still has a closed form inverse. I don't know if it will be useful to other people, but I think it might be...

File: flowjax/bijections/orthogonal.py (hunk 1 above)
Comment (aseyboldt) [reply flowjax/bijections/orthogonal.py:56]: I'm actually not entirely happy with the parametrization of the vector.
It will for instance blow up if `x` is close to zero.
But parameterizing unit vectors isn't trivial, as there is no bijection between any R^n and the space of unit vectors. Maybe we can find something that at least doesn't have any poles?

Comment (aseyboldt) [reply flowjax/bijections/orthogonal.py:56]: Maybe something like this would be a better default parametrization for the unit vector?:

```python

def householder_transform(p, v):
    n = p.shape[0]

    # Compute Householder vector u
    e1 = jnp.zeros_like(p).at[0].set(1.0)
    u = p - e1
    norm_u_raw = jnp.linalg.norm(u)

    # Check if p is already aligned with e1
    is_aligned = norm_u_raw < 1e-6

    # Case 1: If p is approximately e1, embed v as [0, v]
    # This avoids dividing by small values if norm(u) is small
    v_proj_full = jnp.zeros(n).at[1:].set(v)

    # Case 2: Apply the Householder transformation normally
    def apply_householder():
        u_normalized = u / norm_u_raw
        dot_uv = jnp.dot(u_normalized[1:], v)
        v_proj = v - 2.0 * dot_uv * u_normalized[1:]
        return jnp.zeros(n).at[1:].set(v_proj)

    v_proj_full = jnp.where(is_aligned, v_proj_full, apply_householder())

    return v_proj_full


def exp_map_sphere(p, v):
    """Riemannian exponential map on the n-sphere S^n

    Compute the point on the sphere reached by the exponential
    map from point p with tangent vector v (shape: (n+1,))

    Parameters:
    p: Point on the sphere (shape: (n+1,))
    v: Tangent vector in R^n (shape: (n,))
    """
    # Project v into the correct tangent space using Householder transformation
    v_proj = householder_transform(p, v)

    norm_v_raw = jnp.linalg.norm(v_proj)
    # Avoid NaNs in the gradient
    norm_v = jnp.where(norm_v_raw > 1e-6, norm_v_raw, 1.0)

    direction = v_proj / norm_v

    # General case: Compute the exponential map
    exp_general = jnp.cos(norm_v) * p + jnp.sin(norm_v) * direction

    # Small v case: Use a Taylor expansion and re-normalize to stay on the sphere
    exp_taylor = p + v_proj
    exp_taylor = exp_taylor / jnp.linalg.norm(exp_taylor)
    return jnp.where(norm_v_raw > 1e-6, exp_general, exp_taylor)
```

Comment (danielward27) [reply flowjax/bijections/orthogonal.py:56]: I would be hesitant to come up with a custom unit vector parameterization, although I understand the simple approach likely leads to a (probably unavoidable) bad loss landscape. I haven't checked carefully, but most the implementations of related stuff seems to use the simple idea (dividing by the norm for the unit vector or squared norm when computing the householder transform: e.g. [stan](https://github.com/stan-dev/math/blob/develop/stan/math/prim/constraint/unit_vector_constrain.hpp), [original householder flow](https://github.com/jmtomczak/vae_householder_flow/blob/master/models/vae_HF.py#L71-L73), [nflows](https://github.com/bayesiains/nflows/blob/5d22344436a4a5f4c3b93a3b99322a6bfe374c08/nflows/transforms/orthogonal.py#L83)). Division by the norm is pretty common in other applications (e.g. [Weight normalisation](https://arxiv.org/pdf/1602.07868)). I think it makes sense to have the unit vector as a ``Parameterized`` attribute ``self.unit_vec``, so it can be modified as needed. I'll admit I'm no expert here, so I am biased towards the standard approach.

**Comment by danielward27** [comment]
Thanks! ``AsymmetricAffine`` is interesting, but I generally want to try to only include more established methods (to avoid the package becoming cluttered, ensure that people can find full descriptions of the methods, ensure that empirically it provides advantages against other methods etc).

File: flowjax/bijections/orthogonal.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
+from flowjax.bijections.bijection import AbstractBijection
+from jax import Array
+import jax.numpy as jnp
+import jax.nn as jnn
+from jax.scipy import fft
+
+
+class Neg(AbstractBijection):
```
[the remark this answers is not in the document]
Comment (aseyboldt) [reply flowjax/bijections/orthogonal.py]: done

File: flowjax/bijections/orthogonal.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        y_grad: Array,
+        y_logp: Array,
+        condition: Array | None = None,
+    ) -> tuple[Array, Array, Array]:
+        x, logdet = self.inverse_and_log_det(y)
+        x_grad = self._householder(y_grad, params=self.params)
+        return (x, x_grad, y_logp - logdet)
+
+
+class DCT(AbstractBijection):
```
[the remark this answers is not in the document]
Comment (aseyboldt) [reply flowjax/bijections/orthogonal.py]: done

File: flowjax/bijections/orthogonal.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    cond_shape = None
+    axis: int
+    norm: str
+
+    def __init__(self, shape, *, axis: int = -1):
+        self.shape = shape
+        self.axis = axis
+        self.norm = "ortho"
+
+    def _dct(self, x: Array, inverse: bool = False) -> Array:
```
[the remark this answers is not in the document]
Comment (aseyboldt) [reply flowjax/bijections/orthogonal.py]: done

File: flowjax/bijections/utils.py [hunk 5]
```diff
@@ -300,3 +304,54 @@ def inverse_and_log_det(self, y, condition=None):
...
+        inner: Transformation f applied in the middle
+    """
+    shape: tuple[int, ...]
+    cond_shape: tuple[int, ...] | None
+    outer: AbstractBijection
+    inner: AbstractBijection
+
+    def __init__(self, outer: AbstractBijection, inner: AbstractBijection):
+        shape = inner.shape
+        if outer.shape != shape:
```
[the remark this answers is not in the document]
Comment (aseyboldt) [reply flowjax/bijections/utils.py]: Chain for instance also checks in the __init__, and we have to check the cond_shapes anyway, right?
I changed it to use the same helpers as the Chain method.
But if you really don't want it there, I can of course also remove the shape check.

File: flowjax/bijections/orthogonal.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        vec = params / norm
+        return x - 2 * vec * (x @ vec)
+
+    def transform_and_log_det(self, x: jnp.ndarray, condition: Array | None = None):
+        return self._householder(x, self.params), jnp.zeros(())
+
+    def inverse_and_log_det(self, y: Array, condition: Array | None = None):
+        return self._householder(y, self.params), jnp.zeros(())
+
+    def inverse_gradient_and_val(
```
[the remark this answers is not in the document]
Comment (aseyboldt) [reply flowjax/bijections/orthogonal.py]: Wasn't meant to be in there, sorry.

File: flowjax/bijections/orthogonal.py (hunk 1 above)
Comment (aseyboldt) [reply flowjax/bijections/orthogonal.py]: I changed it to the Parameterize call with params / norm(params).

**Review by danielward27 — APPROVED** [review]
Nice! LGTM :) Thanks.
